Sources/Views/AnimatedImageView.swift

```diff
@@ -274,210 +290,262 @@ open class AnimatedImageView: UIImageView {
-    
-    var currentFrame: UIImage? {
-        return frame(at: currentFrameIndexInBuffer)
-    }
+        // The image to display for this frame. Its value is nil when the frame is removed from the buffer.
+        let image: UIImage?
 
-    var isReachMaxRepeatCount: Bool {
-        switch maxRepeatCount {
-        case .once:
-            return currentRepeatCount >= 1
-        case .finite(let maxCount):
-            return currentRepeatCount >= maxCount
-        case .infinite:
-            return false
+        // The duration that this frame should remain active.
+        let duration: TimeInterval
+
+        // A placeholder frame with no image assigned.
+        // Used to replace frames that are no longer needed in the animation.
+        var placeholderFrame: AnimatedFrame {
+            return AnimatedFrame(image: nil, duration: duration)
         }
-    }
-    
-    var contentMode = UIView.ContentMode.scaleToFill
-    
-    private lazy var preloadQueue: DispatchQueue = {
-        return DispatchQueue(label: "com.onevcat.Kingfisher.Animator.preloadQueue")
-    }()
-    
-    /// Creates an animator with image source reference.
-    ///
-    /// - Parameters:
-    ///   - source: The reference of animated image.
-    ///   - mode: Content mode of the `AnimatedImageView`.
-    ///   - size: Size of the `AnimatedImageView`.
-    ///   - count: Count of frames needed to be preloaded.
-    ///   - repeatCount: The repeat count should this animator uses.
-    init(imageSource source: CGImageSource,
-         contentMode mode: UIView.ContentMode,
-         size: CGSize,
-         framePreloadCount count: Int,
-         repeatCount: AnimatedImageView.RepeatCount) {
-        self.imageSource = source
-        self.contentMode = mode
-        self.size = size
-        self.maxFrameCount = count
-        self.maxRepeatCount = repeatCount
-    }
-    
-    func frame(at index: Int) -> Image? {
-        return animatedFrames[safe: index]?.image
-    }
-    
-    func prepareFramesAsynchronously() {
-        preloadQueue.async { [weak self] in
-            self?.prepareFrames()
+
+        // Whether this frame instance contains an image or not.
+        var isPlaceholder: Bool {
+            return image == nil
         }
-    }
-    
-    private func prepareFrames() {
-        frameCount = CGImageSourceGetCount(imageSource)
-        
-        if let properties = CGImageSourceCopyProperties(imageSource, nil),
-           let gifInfo = (properties as NSDictionary)[kCGImagePropertyGIFDictionary as String] as? NSDictionary,
-           let loopCount = gifInfo[kCGImagePropertyGIFLoopCount as String] as? Int
-        {
-            self.loopCount = loopCount
+
+        // Returns a new instance from an optional image.
+        //
+        // - parameter image: An optional `UIImage` instance to be assigned to the new frame.
+        // - returns: An `AnimatedFrame` instance.
+        func makeAnimatedFrame(with newImage: UIImage?) -> AnimatedFrame {
```

Comment:
Factory method's argument label style should be like initializer, use `makeAnimatedFrame(image: UIImage?)` is better.

Reply:
👍